experiment/selinux/src/xattrs/xattr.rs

```diff
@@ -0,0 +1,119 @@
+// lset_xattr sets extended attributes on a symbolic link.
+pub fn lset_xattr(fpath: &Path, attr: &str, data: &[u8]) -> Result<(), XattrError> {
+    rfs::lsetxattr(fpath, attr, data, rfs::XattrFlags::CREATE)
+        .map_err(|e| XattrError::LSetXattr(e.to_string()))
+}
+
+// function similar with getattr in go-selinux repo.
+// get_xattr returns the value of an extended attribute attr set for path.
+pub fn get_xattr(fpath: &Path, attr: &str) -> Result<String, XattrError> {
+    // SELinux label is not so big, so we allocate 1024 bytes for the buffer.
+    let mut buf_size = 1024;
+    let mut buf = vec![0u8; buf_size];
+
+    loop {
+        match rfs::getxattr(fpath, attr, &mut buf) {
+            Ok(size) => {
+                if size == buf_size {
+                    buf_size *= 2;
+                    buf.resize(buf_size, 0);
+                    continue;
+                }
+                let mut value = String::from_utf8_lossy(&buf[..size]).into_owned();
+                if value.ends_with('\x00') {
+                    value.pop();
+                }
+                return Ok(value);
+            },
+            Err(e) => return Err(XattrError::GetXattr(e.to_string())),
+        }
+    }
+}
+
+// function similar with lgetxattr in go-selinux repo.
+// lget_xattr returns the value of an extended attribute attr set for path.
+pub fn lget_xattr(fpath: &Path, attr: &str) -> Result<String, XattrError> {
+    // SELinux label is not so big, so we allocate 1024 bytes for the buffer.
+    let mut buf_size = 1024;
+    let mut buf = vec![0u8; buf_size];
+
+    loop {
+        match rfs::lgetxattr(fpath, attr, &mut buf) {
+            Ok(size) => {
+                if size == buf_size {
+                    buf_size *= 2;
+                    buf.resize(buf_size, 0);
+                    continue;
+                }
+                let mut value = String::from_utf8_lossy(&buf[..size]).into_owned();
+                if value.ends_with('\x00') {
+                    value.pop();
+                }
+                return Ok(value);
+            },
+            Err(e) => return Err(XattrError::LGetXattr(e.to_string())),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::xattrs::*;
+    use std::fs;
+    use std::path::Path;
+    use std::os::unix::fs::symlink;
+    use tempfile::NamedTempFile;
+
+    #[test]
+    fn test_set_xattr_and_get_xattr() {
+        // Because of the permission issue, "selinux.security" can't be used here.
+        let attr_name = "user.test_attr";
+        let attr_value = "system_u:object_r:some_label_t";
+        let temp_file = NamedTempFile::new().expect("Failed to create temp file");
+        let file_path = temp_file.path();
+
+        set_xattr(&file_path, attr_name, attr_value.as_bytes()).expect("Failed to set xattr");
+        let actual = get_xattr(&file_path, attr_name).expect("Failed to get xattr");
+        assert_eq!(actual, attr_value);
+    }
+
+    #[test]
+    #[ignore]
```

Comment:
Hey, if we are ignoring this test, should we add this in the first place? If our original discussion on this was correct, then this will never pass unless run with root permissions, and in this case this does not actually test anything at all when running as non-root.

Reply:
Sure. If we can't run the test with root permissions, I will remove this test.

Reply:
Yes, I think that would be preferable, as in our CI and most places we run tests as non-root

Reply:
Removed them, thanks!